window_manager/src/MsWindow.cpp

```diff
@@ -0,0 +1,177 @@
+:
+config_ (std::move(config))
+, context_(context) {
+	if (!this->create(config_.caption.c_str(), WS_OVERLAPPEDWINDOW, 0, config_.position.x(), config_.position.y(), config_.size.width, config_.size.height))
+		throw std::runtime_error("Can't create window!");
+
+	rendering_context_ = std::make_shared<rcbe::rendering::RenderingContext>();
+
+	rendering_context_->setWindow(getWindowHandle());
+	rendering_context_->setInstance(context_->getInstanceHandle());
+}
+
+MsWindow::~MsWindow() {}
+
+void MsWindow::startWindowLoop() {
+	bool expected = running_.load();
+	running_.compare_exchange_strong(expected, true);
+
+	windowLoop();
+}
+
+void MsWindow::stopWindowLoop() {
+	bool expected = running_.load();
+	running_.compare_exchange_strong(expected, false);
+}
+
+void MsWindow::show() {
+	ShowWindow(getWindowHandle(), context_->getShowCommand());
+}
+
+LRESULT MsWindow::handleMessage(rdmn::core::windows_input_event we) {
+	if (!config_.process_input) return true;
+
+	if(input_manager_) {
+		auto event_processed = input_manager_->tryProcessEvent(we);
+
+		if (event_processed) return true;
+	}
+
+	switch (we.type)
+	{
+	case WM_DESTROY: {
+		PostQuitMessage(0);
+		stopWindowLoop();
+	}
+	return false;
+
+	case WM_PAINT: {
+		PAINTSTRUCT ps;
+		HDC hdc = BeginPaint(window_handle_, &ps);
+		const auto &bg_color = config_.background_color;
+		HBRUSH brush = CreateSolidBrush(RGB(bg_color.r() * 255, bg_color.g() * 255, bg_color.b() * 255)); // TODO: add functionality to RGBAColor
+		FillRect(hdc, &ps.rcPaint, brush);
+
+		/*COLORREF color = RGB(255,255,0);
+		SetConsoleTitle(L"Pixel In Console?");
+		for( int i = 0 ; i < 50 ; i++ )
+		{
+			SetPixel(hdc, 5+i, 12, color); // SetPixel(HDC hdc, int x, int y, COLORREF color)
+		}*/
+
+		EndPaint(window_handle_, &ps);
+		DeleteObject(brush);
+	}
+	return false;
+
+	case WM_COMMAND: {
+		
+	}
+	return false;
+
+	default:
+		return DefWindowProc(window_handle_, we.type, we.wparam, we.lparam);
+	}
+	return true;
+}
+
+LRESULT MsWindow::handleMessage(UINT msg, WPARAM wparam, LPARAM lparam) {
+	switch (msg)
+	{
+	case WM_DESTROY: {
+		PostQuitMessage(0);
+		stopWindowLoop();
+	}
+	return false;
+
+	case WM_PAINT: {
+		PAINTSTRUCT ps;
+		HDC hdc = BeginPaint(window_handle_, &ps);
+		const auto &bg_color = config_.background_color;
+		HBRUSH brush = CreateSolidBrush(RGB(bg_color.r() * 255, bg_color.g() * 255, bg_color.b() * 255)); // TODO: add functionality to RGBAColor
+		FillRect(hdc, &ps.rcPaint, brush);
+		EndPaint(window_handle_, &ps);
+		DeleteObject(brush);
+	}
+	return false;
+
+	case WM_COMMAND: {
+		
+	}
+	return false;
+
+	default:
+		return DefWindowProc(window_handle_, msg, wparam, lparam);
+	}
+	return true;
+}
+
+void MsWindow::windowLoop() {
+	MSG msg { };
+	while (GetMessage(&msg, nullptr, 0, 0) && running_.load())
+	{
+		TranslateMessage(&msg);
+		DispatchMessage(&msg);
+	}
+}
+
+PCWSTR MsWindow::windowClassName() {
+	return L"Windows WinAPI Window";
```

Comment:
This way, all windows will have same class names, is it ok if we say want to have several windows?

Reply:
Not a problem, as far as I understand. One class - several windows.

Reply:
This function is now obsolete.